src/targets/gpu/loop.cpp

```diff
@@ -0,0 +1,134 @@
+#include <iterator>
+#include <migraphx/gpu/loop.hpp>
+#include <migraphx/gpu/context.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace gpu {
+
+shape hip_loop::compute_shape(std::vector<shape> inputs, std::vector<module_ref> mods) const
+{
+    inputs.pop_back();
+    inputs.pop_back();
+    inputs.erase(inputs.begin() + 3);
+    inputs.erase(inputs.begin() + 1);
+    return op.compute_shape(inputs, mods);
+}
+
+static std::pair<int, bool> get_name_index(const std::string& name, const std::string& param_prefix)
+{
+    auto loc = name.find(param_prefix);
+    if(loc != std::string::npos)
+    {
+        int index = std::stoi(name.substr(loc + param_prefix.size()));
+        return {index, true};
+    }
+
+    std::string out_prefix = "#output_";
+    loc                    = name.find(out_prefix);
+    if(loc != std::string::npos)
+    {
+        int index = std::stoi(name.substr(loc + out_prefix.size()));
+        return {index, false};
+    }
+
+    return {-1, false};
+}
+
+argument
+hip_loop::compute(const shape&,
+                  const std::vector<argument>& args,
+                  const std::vector<module_ref>& mods,
+                  const std::function<std::vector<argument>(
+                      module_ref&, const std::unordered_map<std::string, argument>&)>& run) const
+{
+    auto cpy_args = args;
+    std::vector<argument> cpu_args;
+    cpu_args.push_back(cpy_args.at(1));
+    cpu_args.push_back(cpy_args.at(3));
+    cpy_args.erase(cpy_args.begin() + 3);
+    cpy_args.erase(cpy_args.begin() + 1);
+
+    auto iter_num = cpu_args.at(0).at<int64_t>();
+    if(iter_num > op.max_iter_num)
+    {
+        MIGRAPHX_THROW("LOOP compute(): iter_num " + std::to_string(iter_num) +
+                       " larger than max_iter_num " + std::to_string(op.max_iter_num) +
+                       ", please set the max_iter_num to at least " + std::to_string(iter_num) +
+                       " in onnx_options before parsing onnx file!");
+    }
+
+    auto cond                = cpu_args.at(1).at<bool>();
+    module_ref mod           = mods.at(0);
+    auto input_num           = cpy_args.size() - 2;
+    auto dep_num             = input_num - 2;
+    auto param_name_shapes   = mod->get_parameter_shapes();
+    std::string param_prefix = "#" + mod->name() + "_in_";
+    std::vector<argument> in_args(cpy_args.begin(), cpy_args.begin() + input_num);
+    std::vector<argument> out_args = {cpy_args.at(input_num)};
+    auto mod_outputs               = cpy_args.back().get_sub_objects();
+    out_args.insert(out_args.end(), mod_outputs.begin(), mod_outputs.end());
+
+    int64_t iter = 0;
+    for(iter = 0; (iter < iter_num) and cond; ++iter)
+    {
+        // copy iter num and cond to device memory
+        (void)hipMemcpy(in_args.at(0).data(), &iter, sizeof(int64_t), hipMemcpyHostToDevice);
+        (void)hipMemcpy(in_args.at(1).data(), &cond, sizeof(bool), hipMemcpyHostToDevice);
```

Comment:
Use the type safe `copy_to_gpu` and `copy_from_gpu`.

Reply:
And instead of doing `in_args.at(0)` can we assign a name to it such as `gpu_cond_arg = args.at(0)`.